GoRogue/MapGeneration/Region.cs

```diff
@@ -281,46 +302,34 @@ public void RemoveOverlappingPoints(Region imposing)
         /// This is destructive to the region's Connections, so you should try to refrain from generating those
         /// until after you've performed your rotations.
         /// </remarks>
-        public virtual Region Rotate(double degrees, bool doToSelf, Point origin = default)
+        public virtual Region Rotate(double degrees, bool doToSelf, Point origin)
         {
+            degrees = MathHelpers.WrapAround(degrees, 360);
             double radians = SadRogue.Primitives.MathHelpers.ToRadian(degrees);
-
             List<Point> corners = new List<Point>();
-            if (origin == default(Point))
-                origin = new Point((Left + Right) / 2, (Top + Bottom) / 2);
 
-            degrees = MathHelpers.WrapAround(degrees, 360);
 
-            Point sw = SouthWestCorner - origin;
-            int x = (int) Math.Round(sw.X * Math.Cos(radians) - sw.Y * Math.Sin(radians));
-            int y = (int) Math.Round(sw.X * Math.Sin(radians) + sw.Y * Math.Cos(radians));
-            corners.Add(new Point(x, y) + origin);
+            Point sw = RotatePoint(SouthWestCorner - origin, radians) + origin;
+            corners.Add(sw);
 
-            Point se = SouthEastCorner - origin;
-            x = (int) Math.Round(se.X * Math.Cos(radians) - se.Y * Math.Sin(radians));
-            y = (int) Math.Round(se.X * Math.Sin(radians) + se.Y * Math.Cos(radians));
-            corners.Add(new Point(x, y) + origin);
+            Point se = RotatePoint(SouthEastCorner - origin, radians) + origin;
+            corners.Add(se);
 
-            Point nw = NorthWestCorner - origin;
-            x = (int) Math.Round(nw.X * Math.Cos(radians) - nw.Y * Math.Sin(radians));
-            y = (int) Math.Round(nw.X * Math.Sin(radians) + nw.Y * Math.Cos(radians));
-            corners.Add(new Point(x, y) + origin);
+            Point nw = RotatePoint(NorthWestCorner - origin, radians) + origin;
+            corners.Add(nw);
 
-            Point ne = NorthEastCorner - origin;
-            x = (int) Math.Round(ne.X * Math.Cos(radians) - ne.Y * Math.Sin(radians));
-            y = (int) Math.Round(ne.X * Math.Sin(radians) + ne.Y * Math.Cos(radians));
-            corners.Add(new Point(x, y) + origin);
+            Point ne = RotatePoint(NorthEastCorner - origin, radians) + origin;
+            corners.Add(ne);
 
-            sw = corners.OrderBy(c => -c.Y).ThenBy(c => c.X).First();
-            corners.Remove(sw);
+            corners = corners.OrderBy(corner => Direction.YIncreasesUpward ? -corner.Y : corner.Y).ToList();
+            Point[] topTwo = new Point[2] { corners[0], corners[1] };
+            Point[] bottomTwo = new Point[2] {corners [2], corners [3] };
 
-            se = corners.OrderBy(c => -c.Y).First();
-            corners.Remove(se);
+            sw = bottomTwo.OrderBy(c => c.X).ToArray()[0];
+            se = bottomTwo.OrderBy(c => c.X).ToArray()[1];
```

Comment:
Suggested for efficiency to avoid allocating two arrays:
```suggestion
            Point[] bottomTwoOrdered = bottomTwo.OrderBy(c => c.X).ToArray();
            sw = bottomTwoOrdered[0];
            se = bottomTwoOrdered[1];
```